Use dep v0.4.1 in kubernetes/test-infra

ixdy opened the pull request:
It does seem like the new release of dep does a better job of removing unused libraries.

In this PR, I enabled the new prune options (I turned off pruning non-go files, because Bazel).

I then ran `dep ensure` followed by `hack/update-bazel.sh` - I didn't run any of @fejta's other dep scripts in `hack/`. It seems like it removed some dependencies that the scripts missed, but potentially added some it had removed?

Amazingly, `bazel build ... && bazel test ...` still works.

fixes #5488

fejta commented:
IMO we should simplify update-deps.sh (remove the prune command, and the drop-dep calls) and still use it.

ixdy commented:
For fun, I ran `hack/prune-libraries.sh --fix` and added a commit with its results:
```console
$ hack/prune-libraries.sh --fix
Unused libraries:
  DEAD: //vendor/google.golang.org/appengine/urlfetch:go_default_library
  DEAD: //vendor/google.golang.org/appengine/internal/urlfetch:go_default_library
  DEAD: //vendor/google.golang.org/appengine:go_default_library
  DEAD: //vendor/google.golang.org/appengine/internal/modules:go_default_library
  DEAD: //vendor/google.golang.org/appengine/internal/app_identity:go_default_library
  DEAD: //vendor/google.golang.org/appengine/internal:go_default_library
  DEAD: //vendor/google.golang.org/appengine/internal/remote_api:go_default_library
  DEAD: //vendor/google.golang.org/appengine/internal/log:go_default_library
  DEAD: //vendor/google.golang.org/appengine/internal/datastore:go_default_library
  DEAD: //vendor/google.golang.org/appengine/internal/base:go_default_library
  DEAD: //vendor/golang.org/x/text/internal/ucd:go_default_library
  DEAD: //vendor/golang.org/x/text/internal/triegen:go_default_library
  DEAD: //vendor/golang.org/x/text/internal/gen:go_default_library
  DEAD: //vendor/golang.org/x/text/unicode/cldr:go_default_library
  DEAD: //vendor/github.com/spf13/cobra/cobra/cmd:go_default_library
  DEAD: //vendor/github.com/petar/GoLLRB/llrb:go_default_library
Cleaning up unused dependencies...
~/code/kubernetes/src/k8s.io/test-infra ~/code/kubernetes/src/k8s.io/test-infra
~/code/kubernetes/src/k8s.io/test-infra
INFO: Empty results
All remaining go libraries are alive
```

ixdy commented:
And I pushed one more commit which simplifies `hack/update-deps.sh`.

```console
real    0m50.324s
user    0m15.784s
sys     0m5.880s
```

We probably need some way to ensure that the right version of dep is being used.

stevekuznetsov commented:
Simple solution would be to put the script in a presubmit and validate no changes occur?

BenTheElder commented:
/meow box

krzyzacy commented:
/joke

BenTheElder commented:
/meow boxes

fejta commented on hack/update-deps.sh:
```diff
@@ -49,15 +49,7 @@ drop-dep() {
 dep ensure -v
-dep prune -v
 hack/update-bazel.sh
-drop-dep vendor/golang.org/x/text/language vendor/golang.org/x/text/internal
```
So amazing! Drop the `drop-dep` method?

ixdy replied:
done

ixdy commented:
I was able to vendor dep and get it to run under `bazel run`, if we want to ensure that `hack/update-deps.sh` is consistent for everyone.

BenTheElder commented:
fixes  https://github.com/kubernetes/test-infra/issues/5488

BenTheElder commented on hack/update-deps.sh:
```diff
@@ -24,40 +24,20 @@ set -o errexit
+# https://github.com/golang/dep/issues/1580
+rm -rf vendor/github.com/golang/dep/internal/fs/testdata
 hack/update-bazel.sh
-drop-dep vendor/golang.org/x/text/language vendor/golang.org/x/text/internal
```
:tada:

BenTheElder commented:
`/hold` so as to not merge this sort of thing late at night, but this looks good to me. 
We should definitely make sure this runs in CI soon.

ixdy commented:
I thought we were running `hack/verify-deps.sh` somewhere, but I guess not.

/unhold

ixdy commented:
added one more commit, PTAL and unhold when happy

stevekuznetsov commented:
What's the UX for using the vendored `dep` for normal dev work after this change?

BenTheElder commented:
@stevekuznetsov you can use it with something like `bazel run --run_under="cd ${PWD} &&" 
 //vendor/github.com/golang/dep/cmd/dep -- ensure -v` but we shouldn't necessarily need to as we just use hack/update-deps

ixdy commented:
yeah, it's not super elegant. I could maybe add a wrapper script or something.

stevekuznetsov commented:
As long as it's obvious for me when I check in my code that the presubmits are failing because I may have used the wrong `dep` to do my PR.

ixdy commented:
I guess we don't actually have any presubmits checking this, yet. We should figure out how to do that... I'm not sure how to query bazel from inside a test running under bazel, though.

ixdy commented:
I have a somewhat clever way of making `dep` work well under `bazel run`, similar to the approach gazelle used. Will open a PR against repo-infra. (Still not sure about making the verify-deps script run well under bazel.)

stevekuznetsov commented:
So are we generally expecting PR reviewers to notice a huge diff in `vendor/` and enforce that?

stevekuznetsov commented:
/meow space

BenTheElder commented:
Any updates on this? imho we should get this finished and in and follow-up with a presubmit later.

ixdy commented:
waiting on final approval for https://github.com/kubernetes/repo-infra/pull/60.

BenTheElder commented:
Seems I was only looking for confirmation on nits. Maybe we should merge than and nits can be addressed later?

ixdy commented:
OK, this PR is officially ginormous.

I've added several additional commits:
* added a dependency on kubernetes/repo-infra for `workspace_binary`, which means you can now `bazel run //:dep --` and get the right version
* updated the docs around dependencies to mention version, how to use `bazel run`, and the need to run `hack/update-deps.sh` too
* created a simple prow job (copied from verify-config) which runs `hack/verify-deps.sh`, but only when anything deps-related has changed

PTAL

BenTheElder commented:
> OK, this PR is officially ginormous.

We need a new size label? :joy:

BenTheElder commented:
better yet since this is all about vendor and sandboxing:
/meow boxes

ixdy commented:
ok, I'm reasonably sure this works.

/unhold

ixdy commented:
I'll note that the verify job I added here is only running `hack/verify-deps.sh`, which doesn't call dep, just the `hack/prune-libraries.sh` script. I'm not sure if we want to add a `dep ensure` step to this check, much like the kubernetes/kubernetes `verify-deps.sh` check.

ixdy commented:
The verify job appears to be basically doing the right thing - it at least verifies that you've run `hack/prune-libraries.sh`, hopefully through `hack/update-deps.sh`.
